### PR TITLE
Compound parameter improvement : Fix #283

### LINKED ISSUE #283: [ParameterEditor] CompoundParameter improvement
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of the COMET-WEB
- [x] I have searched [open](https://github.com/RHEAGROUP/COMET-WEB-Community-Edition/issues) and [closed](https://github.com/RHEAGROUP/COMET-WEB-Community-Edition/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description

For now, the CompoundParameter provide edition for each component but all in the same way. An improvement could be done by adding the correct Component based on the ParameterType of the CompoundParameterComponent

FYI: Compound ParameterTypes cannot contain other Compound ParameterTypes, so no nesting involved in Popup.

- [x] Refactor Component ParameterType Selector???
- [x] Only show value labels in the Grid, make sure all aligns well. Long texts can be cut off / trimmed.
- [x] Add a button (or something like that) that will create a popup where editing of compound types can be done
- [x] Implement the popup
- [x] Unit tests

### Logs

  - [ ] There are no logs regarding the issue
  - { ] I have attached a log file or included the relevant section in this issue  
  
### System Configuration

- COMET version:
  - [ ] COMET-WEB: 
  - [ ] Other: 
- Web Services:
  - [ ] COMET Web Services version:

### PR DESCRIPTION
**Pull request by nabilatrhea**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/COMET-WEB-Community-Edition/pulls) open
- [x] I have verified that I am following the COMET-WEB [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/COMET-WEB-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description

Fix #283 

- Representation of the right component of ParameterTypes inside CompoundParameterType 
- Refactor ParameterTypesVM

### CODE REVIEW

File: COMETwebapp/Utilities/CompoundComponentSelectedEvent.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        /// </summary>
+        public CompoundParameterTypeEditorViewModel CompoundParameterTypeEditorViewModel { get; set; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CompoundComponentSelectedEvent" /> class.
+        /// </summary>
+        /// <param name="compoundParameterTypeEditorViewModel">The <see cref="CompoundParameterTypeEditorViewModel" /></param>
+        public CompoundComponentSelectedEvent(CompoundParameterTypeEditorViewModel compoundParameterTypeEditorViewModel)
+        {
+            CompoundParameterTypeEditorViewModel = compoundParameterTypeEditorViewModel;
```
Comment (antoineatrhea) [inline COMETwebapp/Utilities/CompoundComponentSelectedEvent.cs:41]: Missing 'this'

File: COMETwebapp/ViewModels/Components/Shared/ParameterEditors/BooleanParameterTypeEditorViewModel.cs [hunk 2]
```diff
@@ -42,7 +42,7 @@ public class BooleanParameterTypeEditorViewModel : ParameterTypeEditorBaseViewMo
         /// <param name="parameterType">the parameter used for this editor view model</param>
         /// <param name="valueSet">the value set asociated to this editor</param>
         /// <param name="isReadOnly">The readonly state</param>
-        public BooleanParameterTypeEditorViewModel(BooleanParameterType parameterType, IValueSet valueSet, bool isReadOnly) : base(parameterType, valueSet, isReadOnly)
+        public BooleanParameterTypeEditorViewModel(BooleanParameterType parameterType, IValueSet valueSet, bool isReadOnly, int compoundIndex = -1) : base(parameterType, valueSet, isReadOnly, compoundIndex)
```
Comment (antoineatrhea) [inline COMETwebapp/ViewModels/Components/Shared/ParameterEditors/BooleanParameterTypeEditorViewModel.cs:45]: General comment: change the compoundIndex to valueArrayIndex with defaut value of 0. The OnParameterValueChanged method could then be modify as : 
```                   
modifiedValueArray = new ValueArray<string>(this.CurrentParameterSwitchKind)
{
    [valueArrayIndex] = valueString
};
```

For component, also you the valueArrayIndex to have the value

File: COMETwebapp/ViewModels/Components/Shared/ParameterEditors/CompoundParameterTypeEditorViewModel.cs [hunk 3]
```diff
@@ -27,23 +27,39 @@ namespace COMETwebapp.ViewModels.Components.Shared.ParameterEditors
     using CDP4Common.EngineeringModelData;
     using CDP4Common.SiteDirectoryData;
     using CDP4Common.Types;
-
+    using CDP4Dal;
     using COMETwebapp.Model;
```
Comment (antoineatrhea) [inline COMETwebapp/ViewModels/Components/Shared/ParameterEditors/CompoundParameterTypeEditorViewModel.cs:31]: missing breakline

File: COMETwebapp/ViewModels/Components/Shared/ParameterEditors/CompoundParameterTypeEditorViewModel.cs [hunk 4]
```diff
@@ -27,23 +27,39 @@ namespace COMETwebapp.ViewModels.Components.Shared.ParameterEditors
     using CDP4Common.EngineeringModelData;
     using CDP4Common.SiteDirectoryData;
     using CDP4Common.Types;
-
+    using CDP4Dal;
     using COMETwebapp.Model;
+    using COMETwebapp.Utilities;
     using COMETwebapp.ViewModels.Components.Viewer.PropertiesPanel;
+    using ReactiveUI;
```
Comment (antoineatrhea) [inline COMETwebapp/ViewModels/Components/Shared/ParameterEditors/CompoundParameterTypeEditorViewModel.cs:34]: missing breakline

File: COMETwebapp/ViewModels/Components/Shared/ParameterEditors/ParameterTypeEditorSelectorViewModel.cs [hunk 5]
```diff
@@ -50,18 +55,29 @@ public class ParameterTypeEditorSelectorViewModel : IParameterTypeEditorSelector
...
 
         /// <summary>
         /// Gets or sets the value set of this <see cref="ParameterType" />
         /// </summary>
         public IValueSet ValueSet { get; set; }
 
+        /// <summary>
+        /// Gets or sets the value set of this <see cref="ParameterType" />
+        /// </summary>
+        public CompoundParameterTypeEditorViewModel SelectedCompoundParameterTypeEditorViewModel { get; set; }
```
Comment (antoineatrhea) [inline COMETwebapp/ViewModels/Components/Shared/ParameterEditors/ParameterTypeEditorSelectorViewModel.cs:79]: Is this property needed?

File: COMETwebapp/ViewModels/Components/Shared/ParameterEditors/ParameterTypeEditorSelectorViewModel.cs [hunk 6]
```diff
@@ -50,18 +55,29 @@ public class ParameterTypeEditorSelectorViewModel : IParameterTypeEditorSelector
         /// <param name="parameterType">the <see cref="ParameterType" /> used for this view model</param>
         /// <param name="valueSet">the value set asociated to the ParameterTypeEditor</param>
         /// <param name="isReadOnly">Value asserting that the <see cref="IParameterEditorBaseViewModel{T}" /> should be readonly</param>
-        public ParameterTypeEditorSelectorViewModel(ParameterType parameterType, IValueSet valueSet, bool isReadOnly)
+        public ParameterTypeEditorSelectorViewModel(ParameterType parameterType, IValueSet valueSet, bool isReadOnly, int compoundIndex = -1, EventCallback<IValueSet>? onParameterValueSetChanged = null)
```
Comment (antoineatrhea) [inline COMETwebapp/ViewModels/Components/Shared/ParameterEditors/ParameterTypeEditorSelectorViewModel.cs:58]: I would not add the EventCallback inside the constructor, please set it from the CompoundParameterTypeEditorVM when creating it

File: COMETwebapp/ViewModels/Components/Shared/ParameterEditors/ParameterTypeEditorSelectorViewModel.cs [hunk 7]
```diff
@@ -50,18 +55,29 @@ public class ParameterTypeEditorSelectorViewModel : IParameterTypeEditorSelector
         /// <param name="parameterType">the <see cref="ParameterType" /> used for this view model</param>
         /// <param name="valueSet">the value set asociated to the ParameterTypeEditor</param>
         /// <param name="isReadOnly">Value asserting that the <see cref="IParameterEditorBaseViewModel{T}" /> should be readonly</param>
```
Comment (antoineatrhea) [inline COMETwebapp/ViewModels/Components/Shared/ParameterEditors/ParameterTypeEditorSelectorViewModel.cs:57]: Missing doccumentation for parameter

**Review by antoineatrhea — CHANGES_REQUESTED** [review]
Please check comments and adjust

**Comment by antoineatrhea** [comment]
@nabilatrhea Please use the ValueArray property instead of this.ValueSet.ActualValue when updating the value

File: COMETwebapp/ViewModels/Components/Shared/ParameterEditors/CompoundParameterTypeEditorViewModel.cs [hunk 8]
```diff
@@ -63,6 +83,15 @@ public override async Task OnParameterValueChanged(object value)
             }
         }
 
+        /// <summary>
+        /// Event for when the edit button is clicked
+        /// </summary>
+        public void OnComponentSelected()
+        {
+            CDPMessageBus.Current.SendMessage(new CompoundComponentSelectedEvent(this));
```
Comment (jaimeatrhea) [inline COMETwebapp/ViewModels/Components/Shared/ParameterEditors/CompoundParameterTypeEditorViewModel.cs:91]: @antoineatrhea Are we using again the MessageBus inside the ParameterEditors?

Comment (antoineatrhea) [reply COMETwebapp/ViewModels/Components/Shared/ParameterEditors/CompoundParameterTypeEditorViewModel.cs:91]: It's only used to open a popup created inside the ParameterTable. Only the ParameterTale subscribed to it. Easier to do it like that.

**Review by antoineatrhea — APPROVED** [review]

**Review by jaimeatrhea — APPROVED** [review]

**Pull request merged**